pxf/pxf-json/src/main/java/org/apache/hawq/pxf/plugins/json/JsonResolver.java

```diff
@@ -0,0 +1,255 @@
+package org.apache.hawq.pxf.plugins.json;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hawq.pxf.api.OneField;
+import org.apache.hawq.pxf.api.OneRow;
+import org.apache.hawq.pxf.api.ReadResolver;
+import org.apache.hawq.pxf.api.io.DataType;
+import org.apache.hawq.pxf.api.utilities.ColumnDescriptor;
+import org.apache.hawq.pxf.api.utilities.InputData;
+import org.apache.hawq.pxf.api.utilities.Plugin;
+import org.codehaus.jackson.JsonFactory;
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.map.ObjectMapper;
+
+/**
+ * This JSON resolver for PXF will decode a given object from the {@link JsonAccessor} into a row for HAWQ. It will
+ * decode this data into a JsonNode and walk the tree for each column. It supports normal value mapping via projections
+ * and JSON array indexing.
+ */
+public class JsonResolver extends Plugin implements ReadResolver {
+
+	private static final Log LOG = LogFactory.getLog(JsonResolver.class);
+
+	private ArrayList<OneField> oneFieldList;
+	private ColumnDescriptorCache[] columnDescriptorCache;
+	private ObjectMapper mapper;
+
+	public JsonResolver(InputData inputData) throws Exception {
+		super(inputData);
+		oneFieldList = new ArrayList<OneField>();
+		mapper = new ObjectMapper(new JsonFactory());
+
+		// Pre-generate all column structure attributes concerning the JSON to column value resolution.
+		columnDescriptorCache = new ColumnDescriptorCache[inputData.getColumns()];
+		for (int i = 0; i < inputData.getColumns(); ++i) {
+			ColumnDescriptor cd = inputData.getColumn(i);
+			columnDescriptorCache[i] = new ColumnDescriptorCache(cd);
+		}
+	}
+
+	@Override
+	public List<OneField> getFields(OneRow row) throws Exception {
+		oneFieldList.clear();
+
+		String jsonRecordAsText = row.getData().toString();
+
+		JsonNode root = decodeLineToJsonNode(jsonRecordAsText);
+
+		if (root == null) {
+			LOG.warn("Return null-fields row due to invalid JSON:" + jsonRecordAsText);
```

Comment:
another possible improvement: create a final list of null fields for the case where the root is null, and immediately return that list here.
That will save us the iteration over the columns and checking for each one if the root is null.